daemon/containerd/image_history.go

```diff
@@ -2,14 +2,104 @@ package containerd
 
 import (
 	"context"
-	"errors"
+	"encoding/json"
 
+	"github.com/containerd/containerd/content"
+	containerdimages "github.com/containerd/containerd/images"
+	cplatforms "github.com/containerd/containerd/platforms"
+	"github.com/docker/distribution/reference"
 	imagetype "github.com/docker/docker/api/types/image"
-	"github.com/docker/docker/errdefs"
+	"github.com/docker/docker/pkg/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
-// ImageHistory returns a slice of ImageHistory structures for the specified
-// image name by walking the image lineage.
+// ImageHistory returns a slice of HistoryResponseItem structures for the
+// specified image name by walking the image lineage.
 func (i *ImageService) ImageHistory(ctx context.Context, name string) ([]*imagetype.HistoryResponseItem, error) {
-	return nil, errdefs.NotImplemented(errors.New("not implemented"))
+	desc, err := i.resolveDescriptor(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	cs := i.client.ContentStore()
+	// TODO: pass the platform from the cli
+	conf, err := containerdimages.Config(ctx, cs, desc, platforms.AllPlatformsWithPreference(cplatforms.Default()))
+	if err != nil {
+		return nil, err
+	}
+
+	diffIDs, err := containerdimages.RootFS(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	blob, err := content.ReadBlob(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	var image ocispec.Image
+	if err := json.Unmarshal(blob, &image); err != nil {
+		return nil, err
+	}
+
+	var (
+		history []*imagetype.HistoryResponseItem
+		sizes   []int64
+	)
+	s := i.client.SnapshotService(i.snapshotter)
+
+	for i := range diffIDs {
+		chainID := identity.ChainID(diffIDs[0 : i+1]).String()
+
+		use, err := s.Usage(ctx, chainID)
+		if err != nil {
+			return nil, err
+		}
+
+		sizes = append(sizes, use.Size)
+	}
```

Comment:
"Size" of images in containerd is going to continue to be an interesting concept to grapple with, isn't it? :joy:

(Size of the content, for which platform, for which snapshotter, etc etc etc forever :sob:)